File: LeetCode_Algorithm/917(Reverse Only Letters)/917.py
'''
LeetCode 917 : Reverse Only Letters
주어진 문자열이 있을 때 규칙에 따라 문자를 뒤집어라.
영어가 아닌 경우 문자는 그 위치에 그대로 남아야 한다.
영어는 뒤집어야 한다.

# sol1 32ms 74% 14MB 11%
ans를 초기화 하고 영어가 아닌 문자들을 먼저 배치 한 뒤 
영어를 뒤집어서 넣음

# sol2 22ms 98% 13.9MB 58%
two stack 
두 개의 스택에 저장해 놓고 영어는 거꾸로 넣고 글자는 그냥 넣기
'''


class Solution:
    def reverseOnlyLetters(self, s: str) -> str:
        # sol1
        # idx = 0
        # ans = [' ' for _ in range(len(s))]
        # while idx < len(s):
        #     if not s[idx].isalpha():
        #         ans[idx] = s[idx]
        #     idx += 1
        # ansIdx = 0; idx = len(s)-1
        # while True:
        #     while ansIdx < len(ans) and ans[ansIdx] != ' ':
        #         ansIdx += 1
        #     while idx >= 0 and not s[idx].isalpha():
        #         idx -= 1
        #     if ansIdx < len(ans) and idx >= 0:
        #         ans[ansIdx] = s[idx]
        #         ansIdx += 1
        #         idx -= 1
        #     else:
        #         break
        # return "".join(ans)

        # sol2
        # eng = []
        # others = []
        # ans = []
        # for c in s:
        #     if c.isalpha():
        #         eng.append(c)
        #     else:
        #         others.append(c)
        # for i in range(len(s)):
        #     if not s[i].isalpha():
        #         c = others.pop(0)
        #         ans.append(c)
        #     else:
        #         c = eng.pop()
        #         ans.append(c)
        # return "".join(ans)

        # sol3 33ms 86% 16.09MB 98%
        # O(n) O(n) 
        ans = ""
        i, j = 0, len(s)-1
        while i < len(s):
            if s[i].isalpha():
                while j > 0 and not s[j].isalpha():
                    j -= 1
                ans += s[j]
                j -= 1
            else:
                ans += s[i]
            i += 1
        return ans
                
    
print(Solution().reverseOnlyLetters("?6C40E"))
print(Solution().reverseOnlyLetters("----"))
print(Solution().reverseOnlyLetters("babaab"))
print(Solution().reverseOnlyLetters("dc-ba"))
print(Solution().reverseOnlyLetters("a-bC-dEf-ghIj"))
print(Solution().reverseOnlyLetters("Test1ng-Leet=code-Q!"))





        
'''
?6E40C
?6E4C0

?6C40E
?6E40
'''

File: LeetCode_Algorithm/917(Reverse Only Letters)/917.js
var isalpha = (c) => {
  if (c >= 'a' && c <= 'z') 
    return true
  if (c >= 'A' && c <= 'Z')
    return true
  return false
}
var reverseOnlyLetters = function(s) {
  // sol2 68ms 61% 42.3MB 47%
  // let eng = []; let others = []
  // let ans = []
  // for (let c of s) {
  //   if (isalpha(c))
  //     eng.push(c)
  //   else
  //     others.push(c)
  // }
  // for (let i =0; i <s.length; i++) {
  //   if (isalpha(s[i])) {
  //     let c = eng.pop()
  //     ans.push(c)
  //   } else {
  //     let c = others.shift()
  //     ans.push(c)
  //   }
  // }
  // return ans.join("")

  // sol3 51ms 63% 42.29MB 44MB
  let ans = ""
  let i = 0; let j = s.length-1
  while (i < s.length) {
    if (!isalpha(s[i])) {
      ans += s[i]
    } else {
      while (j > 0 && !isalpha(s[j])) {
        j -= 1
      }
      ans += s[j]
      j -= 1
    }
    i += 1
  }
  return ans
};

console.log(reverseOnlyLetters("ab-cd"))